design/automatic-network-configuration.md

```diff
@@ -0,0 +1,801 @@
+convenient to be able to reconfigure related network devices using a
+similar API.
+
+## Goals
+
+- Define a Kubernetes API for configuring network device.
+- Automatically configure the network infrastructure for a host when
+adding it to a cluster.
+- Do network configuration when deprovisioning the hosts.
+- Design a network abstraction that can represent any of the target networking
+configuration, independently of the controller used.
+
+## Non-Goals
+
+- Implement a network controller which is aimed at being an integration with
+an existing controller only.
+- Implement a solution for a specific underlying infrastructure.
+- Implement a solution for link aggregation in the future.
+- Implement a solution for lldp in the future.
+
+## Proposal
+
+This document proposes to add a new mechanic to automatically perform physical
+network device configuration before provisioning a BareMetalHost. First we
+added the NetworkConfiguration operator to implement the network configuration
+of the device. Then we added some methods to CAPM3 to implement together. When
+network configuration is required, CAPM3 will process some content through
+additional methods, and then trigger the NetworkConfiguration operator to
+perform network configuration.
+
+### User Stories
+
+#### Story 1
+
+As a consumer of Metal³, when adding a machine resource to the cluster, according
+to the network configuration of the cluster, the physical host is connected to the
+corresponding logical layer-2 network segment.
+
+#### Story 2
+
+As a consumer of Metal³, when a machine is deleted from the cluster, the
+corresponding network configuration is automatically deleted.
+
+#### Story 3
+
+As a user/admin a SmartNIC that is hosted in a BMH needs to be provisioned,
+pre-configured, and then capable of in operation management.
+
+Physical server(OS) -----> SmartNIC [virtual switch(OVS) ---> physical NIC]
+←-- (LAN cable)-----> PortID
+
+#### Story 4
+
+As a user, when creating a network configuration and specifying link
+aggregation, the related BMH and switch can be automatically configured.
+
+#### Story 5
+
+As a user, when creating a network configuration, it is possible to configure ACL
+settings in a switch for each bare metal server according to its functionality.
+The ACL setting:
+
+- Direction: ingress or egress
+- Protocol: SSH, HTTP, …all
+- Action: allow, deny
+- Port status: up, down (status of the switch port)
+
+## Design Details
+
+### Structure
+
+In the network operator, we abstract the following roles:
+
+#### Device
+
+Abstract all devices in the cluster (such as Switch, SmartNic).
+Each type of network device must have its own CRD.
+
+At the same time, each device has its own specific interface, and
+needs to implement corresponding interfaces for different vendors
+and device types. The following is an example of a switch interface:
+```go
+// Switch is a interface for different protocol
+type Switch interface {
+    // PowerOn enable switch
```

Comment:
I don't think we should manage power of switches. Also "enable" and "power on" are different things in bare metal world. Could you clarify what these 2 calls do and also consider if they're needed at all, especially for the first version.

Reply:
updated.